torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,292 @@
+            and only after _all_ gradients for that optimizer's owned parameters
+            have been accumulated.
+        """
+        if not self._enabled:
+            return
+
+        optimizer_state = self._per_optimizer_states[id(optimizer)]
+
+        if "unscaled" in optimizer_state:
+            raise RuntimeError("unscale() has already been called on this optimizer this iteration.")
+
+        # FP32 division can be imprecise for certain compile options, so we carry out the reciprocal in FP64.
+        rscale = self._scale.double().reciprocal().float()
+        found_inf = torch.full((1,), 0.0, dtype=torch.float32, device="cuda")
+
+        self._unscale_grads(optimizer, rscale, found_inf, allow_fp16=False)
+
+        optimizer_state["found_inf"] = found_inf
+        optimizer_state["unscaled"] = True
+
+    def step(self, optimizer, *args, **kwargs):
+        """
+        Carry out a scaling-safe step using ``optimizer``.  "Scaling-safe" means two things:
+
+        1.  If :meth:`unscale` has not yet been invoked, :meth:`step` will make sure to invoke :meth:`unscale`
+            internally.
+        2.  If inf/NaN gradients are found, :meth:`step` will skip ``optimizer.step()`` to avoid polluting the
+            params.
+
+        ``*args`` and ``**kwargs`` will be forwarded to ``optimizer.step()``.
+
+        Arguments:
+            optimizer (torch.optim.Optimizer):  Optimizer that applies the gradients.
+            args:  Any arguments.
+            kwargs:  Any keyword arguments.
+
+        Returns:
+            The return value of ``optimizer.step(*args, **kwargs)``.
+
+        .. note::
+            If you're writing a custom optimizer, and wish to define your own scaling-safe ``step`` method that
+            :meth:`AmpScaler.step` may call directly without any wrapping logic, see the
+            :ref:`Custom Optimizer Guide<custom-optimizer-guide>`.
+
+        .. warning::
+            Closure use is not currently supported.
+        """
+        if (not self._enabled):
+            return optimizer.step(*args, **kwargs)
+
+        if "closure" in kwargs:
+            raise RuntimeError("Closure use is not currently supported if AmpScaler is enabled.")
+
+        if (hasattr(optimizer, "step_supports_amp_scaling") and optimizer.step_supports_amp_scaling):
+            # This optimizer has customized scaling-safe step logic, so we call it directly.
+            # The contract with custom optimizers is that their step methods should accept an additional,
+            # optional amp_scaler kwarg.  We append self to the kwargs so the custom optimizer has full information:
+            # it can query its own state, invoke unscale on its own gradients for convenience, etc.
+            return optimizer.step(*args, **dict(kwargs, amp_scaler=self))
+
+        optimizer_state = self._per_optimizer_states[id(optimizer)]
+
+        if "unscaled" not in optimizer_state:
```

Comment:
You're querying if the key is NOT IN the dict here, but below you're setting it to true, and I'm not sure where it's cleared. Can you describe this stateful behavior?

Reply:
It's cleared in unscale() to prepare for the next iteration.